src/k2/persistence/plog_client/PlogClient.cpp

```diff
@@ -0,0 +1,182 @@
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#include "PlogClient.h"
+#include <k2/common/Chrono.h>
+#include <k2/config/Config.h>
+#include <k2/dto/Collection.h>
+#include <k2/dto/Plog.h>
+#include <k2/transport/RPCDispatcher.h>
+#include <k2/transport/RPCTypes.h>
+#include <k2/transport/Status.h>
+#include <k2/transport/TXEndpoint.h>
+#include <k2/dto/ControlPlaneOracle.h>
+#include <k2/dto/MessageVerbs.h>
+#include <cstdlib>
+#include <cctype>
+#include <algorithm>
+
+namespace k2 {
+
+char Rand_alnum()
+{
+    char c;
+    while (!std::isdigit(c = static_cast<char>(std::rand())));
+    return c;
+}
+
+PlogClient::PlogClient() {
+    K2INFO("dtor");
+}
+
+PlogClient::~PlogClient() {
+    K2INFO("~dtor");
+}
+
+void PlogClient::GetPlogServerEndpoint() {
+    for(auto& v : _partitionCluster.partitionGroupVector){
+        K2INFO("Partition Group: " << v.name);
+        _partitionNameList.push_back(v.name);
+        
+        std::vector<std::unique_ptr<TXEndpoint>> endpoints;
+        for (auto& url: v.plogServerEndpoints){
+            K2INFO("Plog Server Url: " << url);
+            auto ep = RPC().getTXEndpoint(url);
+            endpoints.push_back(std::move(ep));
+        }
+
+        _partitionMapEndpoints[std::move(v.name)] = std::move(endpoints);
+    }
+}
+
+seastar::future<> 
+PlogClient::GetPartitionCluster(String name){
+    _cpo = CPOClient(String(_cpo_url()));
+    return _cpo.GetPartitionCluster(Deadline<>(get_plog_server_deadline()), std::move(name)).
+    then([this] (auto&& result) {
+        auto& [status, response] = result;
+
+        if (!status.is2xxOK()) {
+            K2INFO("Failed to obtain Partition Cluster" << status);
+            return seastar::make_exception_future<>(std::runtime_error("Failed to obtain Partition Cluster Map"));
+        }
+
+        _partitionCluster = std::move(response.cluster);
+        _partition_map_pointer = rand() % _partitionCluster.partitionGroupVector.size();
+        _partitionMapEndpoints.clear();
+        GetPlogServerEndpoint();

```

Comment:
this method should return a future based on whether it succeeded or not. Then here you'd just: `return getPlogServerEndpoint()`

Reply:
fixed